keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,428 @@
+1. `labels`: should be added by controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+1. `finalizers`: should be added by controller to defer `BucketAccess` deletion until backend deletion ops succeed.
+1. `bucketAccessRequestName`/`bucketAccessRequestNamespace`: name and namespace of the bound `BucketAccessRequest`
+1. `serviceAccountName`: name of the Kubernetes ServiceAccount specified by the `BucketAccessRequest`.  Undefined when the `BucketAccessRequest.accessSecretName` is defined.
+1. `  accessSecretName`: name of the *provisioner* generated `Secret` containing access credentials. This `Secret` exists in the provisioner’s namespace and must be copied to the app namespace by the COSI controller.
+1. `provisioner`:  name of the provisioner that should handle this `BucketAccess` instance.  Copied from the `BucketAccessClass`.
+1. `parameters`:  A map of string, string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
+1. `message`: a human readable description detailing the reason for the current `phase``.
+1. `phase`: is the current state of the Bucket:
+   - _Bound_: the controller finished processing the request and has bound the BucketAccess to the BucketAccessRequest
+   - _Released_: the originating Bucket has been deleted, signalling that the Bucket is ready for garbage collection.  This will occur on Greenfield Buckets once all requests referencing the Bucket are deleted.
+   - _Failed_: error and all retries have been exhausted.
+   - _Retrying_: set when a driver or Kubernetes error is encountered during provisioning operations indicating a retry loop.
+
+#### BucketAccessClass
+
+A cluster-scoped API providing a way for admins to specify policies that may be used to access buckets.  These are always applicable in Greenfield, where access is dynamically granted, and only sometimes applicable in Brownfield, where a user identity may already exist, but not yet have access to a bucket.  In this case, a BucketAccessRequest will still specify the BucketAccessClass in order to determine which actions it is granted to a defined bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessClass
+metadata: 
+  name:
+provisioner: [1]
+supportedProtocols: [2]
+policyActions: [3]
+  allow:
+  - "*"
+  deny:
+  - "*"
+parameters: [4]
+```
+
+1. `provisioner`: The name of the provisioner that `BucketAccess` instances should be managed by.
+1. `supportedProtocols`: protocols the associated object store supports.  Applied when matching Bucket to BucketClasses.  Admins may specify more than protocol when necessary.  `BucketAccessRequests.spec.protocol` will be checked against this array prior to provisioning.
```

Comment:
"Applied when matching BucketAccessRequests to BucketAccessClasses."
"Admins may specify more than one protocol when necessary. "

Reply:
"Applied when matching BucketAccessRequests to BucketAccessClasses."
"Admins may specify more than one protocol when necessary. "

Reply:
Multiple default classes are allowed, as long as they are for different protocols. At the same time, a single class can specify multiple protocols. Isn't that confusing? Wouldn't it be easier to have a 1:1 mapping between class and protocol? Additional parameters will likely be protocol-dependent as well, so I'm not sure if mixing multiple protocols in the same class is a good idea.

Reply:
@wonderland GCP and Azure support >1 protocol. The idea is that the BC represents capabilities of the underlying object store and a BR (bucket request) knows which protocol to request. COSI will validate that the referenced BC supports the desired protocol.
Regarding multiple default BCs: our thinking was that an admin could define one default BC per protocol and COSI will find the correct default BC by matching BR's protocol with supported protocols in all default BCs (of which there should not be more than 3). If default BC-1 supports "s3" and default BC-2 supports "gcp, s3" then we have undefined behavior when a BR requests s3. 
We could state that default BCs can only list one supported protocol to make it less confusing. Thoughts?

Reply:
Many on-premises solutions also support several protocols (S3, Swift, ...), but each protocol would typically come with slightly different parameters and capabilities. So mixing them in a single BC, possibly making exceptions to that rule for default BC, seems confusing to me. I would find it more logical to create (at least) one BC for each protocol the underlying object store supports (if you want to use all protocols). In most cases it will probably be just a single protocol anyway, so this wouldn't lead to an inflation of BC on a cluster. But maybe that's just me getting confused easily ;-)

Reply:
@wonderland I agree it's probably less confusing for a default BC to support only 1 protocol, especially when considering _parameters_.  I am fine with this. @wlan0 ?

Reply:
@wonderland this discussion about default classes should be under `BucketClass` not `BucketAccessClass`.